llvm/test/CodeGen/RISCV/GlobalISel/rvv/splat-vector.ll

```diff
@@ -67,7 +67,428 @@ define <vscale x 64 x i1> @splat_zero_nxv64i1() {
   ret <vscale x 64 x i1> zeroinitializer
 }
 
+define <vscale x 1 x i8> @splat_zero_nxv1i8() {
```

Comment:
Add tests splatting a scalar

Reply:
We are not there yet. We need `G_EXTRACT_VECTOR_ELT` for insertelement support.

Reply:
I think that means we're missing a combine to delete a nop G_INSERT_VECTOR_ELT+ G_EXTRACT_VECTOR_ELT pair?

Reply:
Thats correct.

CC: @jiahanxie353 (we came across this)

Reply:
Do you mean:
```
def insert_vector_element_extract_vector_element : GICombineRule<
   (defs root:$root),
   (match (G_EXTRACT_VECTOR_ELT $elt, $src, $idx),
          (G_INSERT_VECTOR_ELT $root, $src, $elt, $idx)),
   (apply (GIReplaceReg $root, $src))>;
```
?

Reply:
I looked at the code coming into the seslector, and it looks like this did not get applied. I'll have to take a closer look.

Reply:
That appears to be matching an insert of an extract. Not an extract from an insert.

Reply:
The splat pattern in IR is
```
insertelt
shufflevector
```

For scalable vectors, the shufflevector becomes extractelt followed by splat_vector so we get

```
insertelt
extractelt
splatvector
```

We need to combine away the extractelt and insertelt.

Reply:
So it should be this combine

```
def extract_vector_element_not_const : GICombineRule<                            
   (defs root:$root),                                                            
   (match (G_INSERT_VECTOR_ELT $src, $x, $value, $idx),                          
          (G_EXTRACT_VECTOR_ELT $root, $src, $idx)),                             
   (apply (GIReplaceReg $root, $value))>
```

Reply:
```
def extract_vector_element_not_const : GICombineRule<
   (defs root:$root),
   (match (G_INSERT_VECTOR_ELT $src, $x, $value, $idx),
          (G_EXTRACT_VECTOR_ELT $root, $src, $idx)),
   (apply (GIReplaceReg $root, $value))>;

def extract_vector_element_different_indices : GICombineRule<
   (defs root:$root, build_fn_matchinfo:$matchinfo),
   (match (G_INSERT_VECTOR_ELT $src, $x, $value, $idx2),
          (G_EXTRACT_VECTOR_ELT $root, $src, $idx1),
   [{ return Helper.matchExtractVectorElementWithDifferentIndices(${root}, ${matchinfo}); }]),
   (apply [{ Helper.applyBuildFnMO(${root}, ${matchinfo}); }])>;
```

Reply:
Which is failing in my local test because the indices weren't CSEd

  %4:_(s64) = G_CONSTANT i64 0                                                                                   
  %2:_(<vscale x 2 x s32>) = G_INSERT_VECTOR_ELT %3:_, %0:_(s32), %4:_(s64)                                      
  %5:_(s64) = G_CONSTANT i64 0                                                                                   
  %6:_(s32) = G_EXTRACT_VECTOR_ELT %2:_(<vscale x 2 x s32>), %5:_(s64)

Reply:
You are really fast.